File: packages/java/tests/TypesafeMCPTest.java
import dev.axllm.ax.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.*;

public final class TypesafeMCPTest {
  static void check(boolean value,String message){if(!value)throw new AssertionError(message);}
  static final class Socket implements AxMCPWebSocketTransport.Socket {
    final BlockingQueue<String> inbound=new LinkedBlockingQueue<>(),sent=new LinkedBlockingQueue<>();
    volatile boolean fail;
    public void send(String text){if(fail)throw new IllegalStateException("send failed");sent.add(text);}
    public String receive()throws Exception{String text=inbound.take();return "CLOSE".equals(text)?null:text;}
    public void close(){inbound.add("CLOSE");}
    void sent()throws Exception{check(sent.poll(2,TimeUnit.SECONDS)!=null,"request did not start");}
    void respond(String value){inbound.add("{\"id\":1,\"result\":\""+value+"\"}");}
  }
  static void error(Future<?> task,String message)throws Exception{
    try{task.get(2,TimeUnit.SECONDS);throw new AssertionError("expected "+message);}catch(ExecutionException e){check(e.getCause().toString().contains(message),e.toString());}
  }
  static void websocket()throws Exception{
    Socket socket=new Socket();var transport=new AxMCPWebSocketTransport("ws://example.test",List.of(),(url,protocols)->socket);transport.setProtocolVersion("2025-03-26");
    var pool=Executors.newFixedThreadPool(3);Map<String,Object> request=Map.of("id",1,"method","ping");var batch=List.of(request,Map.<String,Object>of("id",2,"method","ping"));
    try{
      socket.fail=true;error(pool.submit(()->transport.send(request)),"send failed");error(pool.submit(()->transport.sendBatch(batch)),"send failed");socket.fail=false;
      var old=new AtomicBoolean();var first=pool.submit(()->transport.sendWithContext(request,Map.of(),old::get));socket.sent();socket.respond("first");check(first.get(2,TimeUnit.SECONDS).get("result").equals("first"),"first response");
      var second=pool.submit(()->transport.send(request));socket.sent();old.set(true);
      error(pool.submit(()->transport.send(request)),"already pending");socket.respond("second");check(second.get(2,TimeUnit.SECONDS).get("result").equals("second"),"old cancellation removed reused ID");
      var abort=new AtomicBoolean();var cancelled=pool.submit(()->transport.sendBatch(batch,abort::get));socket.sent();abort.set(true);error(cancelled,"cancelled");
      var closed=pool.submit(()->transport.sendBatch(batch));socket.sent();transport.close();error(closed,"closed");
    }finally{transport.close();pool.shutdownNow();check(pool.awaitTermination(2,TimeUnit.SECONDS),"request workers leaked");}
  }
  @SuppressWarnings("unchecked") static void nativeClient()throws Exception{
    var discovery=new AtomicInteger();var credentials=new AtomicInteger();
    OpenAICompatibleClient.Transport transport=request->{
      if(request.get("method").equals("GET")){
        check(!request.containsKey("json")&&!request.containsKey("data"),"GET has body");
        if(discovery.incrementAndGet()==1)return Map.of("status",429,"json",Map.of("error","retry"));
        return Map.of("models",List.of(Map.of("name","jev-latest","description","Jev","release_date","2026-09-01")));
      }
      var payload=(Map<String,Object>)request.get("json");var answers=new LinkedHashMap<String,Object>();
      for(String key:((Map<String,Object>)payload.get("questions")).keySet())answers.put(key,Map.of("type","noul","noul",((Map<String,Object>)payload.get("state")).get("probability")));
      return Map.of("model",payload.get("model"),"answers",answers,"usage",Map.of("input_tokens",1,"output_tokens",1));
    };
    OpenAICompatibleClient.CredentialProvider credential=request->{credentials.incrementAndGet();return Map.of("Authorization","Bearer test");};
    var client=Ax.typesafe(Map.of("credential_provider",credential,"transport",transport,"retry",Map.of("maxRetries",1,"initialDelayMs",1)));
    check(client.listModels().get(0).name().equals("jev-latest")&&discovery.get()==2&&credentials.get()==2,"model discovery and retry");
    var pool=Executors.newFixedThreadPool(4);
    try{
      List<Future<?>> tasks=new ArrayList<>();
      for(int i=0;i<10;i++){final int index=i;tasks.add(pool.submit(()->{String key="question"+index;double probability=index/10.0;try{
        var response=client.systemOne(Map.of("state",Map.of("probability",probability),"questions",Map.of(key,Map.of("type","noul"))));
        var answers=response.answers();check(((AxAITypesafeClient.Noul)answers.get(key)).noul()==probability,"native probability or question key changed");
      }catch(Exception error){throw new RuntimeException(error);}}));}
      for(var task:tasks)task.get(2,TimeUnit.SECONDS);
      AxCancellationToken token=new AxCancellationToken();token.cancel("cancel native");
      int before=credentials.get();try{client.systemOne(Map.of("state",Map.of("probability",0.5),"questions",Map.of("flag",Map.of("type","noul"))),Map.of("cancellation",token));throw new AssertionError("pre-aborted request accepted");}catch(AxAIServiceAbortedError expected){}
      check(credentials.get()==before,"aborted request accessed credentials");
    }finally{pool.shutdownNow();}
  }
  @SuppressWarnings("unchecked") static void nestedValidation() {
    var typed=Ax.ai("typesafe",Map.of("api_key","test","models",List.of()));
    var only=new AxBalancer(List.of(typed));
    var request=Map.<String,Object>of("chat_prompt",List.of(Map.of("role","user","content","reply")));
    try {only.validateChatRequest(request);throw new AssertionError("nested Typesafe accepted prose");} catch(IllegalArgumentException expected) {}
    only.validateChatRequest((Map<String,Object>)Json.parse("{\"chat_prompt\":[{\"role\":\"user\",\"content\":\"outage\"}],\"response_format\":{\"type\":\"json_schema\",\"schema\":{\"name\":\"decision\",\"schema\":{\"type\":\"object\",\"properties\":{\"urgent\":{\"type\":\"boolean\"}},\"required\":[\"urgent\"]}}}}"));
    var mixed=new AxBalancer(List.of(only,Ax.ai("openai",Map.of("api_key","test","models",List.of()))));
    mixed.validateChatRequest(request);
  }
  public static void main(String[] args)throws Exception{websocket();nativeClient();nestedValidation();System.out.println("Java Typesafe native client and MCP WebSocket cleanup passed");}
}

File: packages/go/typesafe_mcp_test.go
package axllm

import (
	"context"
	"encoding/json"
	"fmt"
	"github.com/coder/websocket"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

type paritySocket struct {
	incoming chan string
	sent     chan string
	closed   chan struct{}
	once     sync.Once
	fail     atomic.Bool
}

func newParitySocket() *paritySocket {
	return &paritySocket{incoming: make(chan string, 8), sent: make(chan string, 8), closed: make(chan struct{})}
}
func (s *paritySocket) Send(text string) error {
	if s.fail.Load() {
		return fmt.Errorf("send failed")
	}
	s.sent <- text
	return nil
}
func (s *paritySocket) Receive() (string, error) {
	select {
	case text := <-s.incoming:
		return text, nil
	case <-s.closed:
		return "", fmt.Errorf("socket closed")
	}
}
func (s *paritySocket) Close() error { s.once.Do(func() { close(s.closed) }); return nil }
func parityWaitSent(t *testing.T, s *paritySocket) {
	t.Helper()
	select {
	case <-s.sent:
	case <-time.After(time.Second):
		t.Fatal("request did not start")
	}
}
func parityWaitResult(t *testing.T, done chan error, want string) {
	t.Helper()
	select {
	case err := <-done:
		if err == nil || !strings.Contains(err.Error(), want) {
			t.Fatalf("expected %q, got %v", want, err)
		}
	case <-time.After(time.Second):
		t.Fatal("pending request leaked")
	}
}
func TestMCPWebSocketCleanup(t *testing.T) {
	socket := newParitySocket()
	transport := NewAxMCPWebSocketTransport("ws://example.test", map[string]Value{"webSocketFactory": AxMCPWebSocketFactory(func(string, []string) (AxMCPWebSocket, error) { return socket, nil })})
	defer transport.Close()
	transport.SetProtocolVersion("2025-03-26")
	request := map[string]Value{"jsonrpc": "2.0", "id": 1, "method": "ping"}
	socket.fail.Store(true)
	if _, err := transport.Send(request); err == nil {
		t.Fatal("missing synchronous send failure")
	}
	if _, err := transport.SendBatch(context.Background(), []Value{request, map[string]Value{"id": 2, "method": "ping"}}); err == nil {
		t.Fatal("missing batch failure")
	}
	if len(transport.pending) != 0 {
		t.Fatal("send failure retained pending work")
	}
	socket.fail.Store(false)
	old, cancelOld := context.WithCancel(context.Background())
	defer cancelOld()
	done := make(chan error, 1)
	go func() { _, err := transport.SendWithContext(old, request, nil); done <- err }()
	parityWaitSent(t, socket)
	socket.incoming <- `{"id":1,"result":"first"}`
	if err := <-done; err != nil {
		t.Fatal(err)
	}
	go func() {
		result, err := transport.Send(request)
		if err == nil && result["result"] != "second" {
			err = fmt.Errorf("wrong response")
		}
		done <- err
	}()
	parityWaitSent(t, socket)
	cancelOld()
	if _, err := transport.Send(request); err == nil {
		t.Fatal("overwrote pending request")
	}
	socket.incoming <- `{"id":1,"result":"second"}`
	if err := <-done; err != nil {
		t.Fatal(err)
	}
	ctx, cancel := context.WithCancel(context.Background())
	go func() {
		_, err := transport.SendBatch(ctx, []Value{request, map[string]Value{"id": 2, "method": "ping"}})
		done <- err
	}()
	parityWaitSent(t, socket)
	cancel()
	parityWaitResult(t, done, "canceled")
	go func() {
		_, err := transport.SendBatch(context.Background(), []Value{request, map[string]Value{"id": 2, "method": "ping"}})
		done <- err
	}()
	parityWaitSent(t, socket)
	_ = transport.Close()
	parityWaitResult(t, done, "closed")
	transport.mu.Lock()
	defer transport.mu.Unlock()
	if len(transport.pending) != 0 {
		t.Fatal("close retained pending work")
	}
}
func TestTypesafeNativeHTTP(t *testing.T) {
	var calls atomic.Int32
	started := make(chan struct{}, 1)
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/v1/models" {
			if r.Method != "GET" || r.ContentLength != 0 {
				t.Error("model discovery must use bodyless GET")
			}
			if r.Header.Get("Authorization") != "Bearer test" {
				t.Error("missing credentials")
			}
			if calls.Add(1) == 1 {
				w.WriteHeader(429)
				_, _ = w.Write([]byte(`{"error":"retry"}`))
				return
			}
			_, _ = w.Write([]byte(`{"models":[{"name":"jev-latest","description":"Jev","release_date":"2026-09-01"}]}`))
			return
		}
		var body map[string]Value
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Error(err)
			return
		}
		if body["state"] == "stall" {
			w.Header().Set("Content-Length", "10000")
			w.WriteHeader(200)
			w.(http.Flusher).Flush()
			started <- struct{}{}
			<-r.Context().Done()
			return
		}
		answers := map[string]Value{}
		for key := range body["questions"].(map[string]Value) {
			answers[key] = map[string]Value{"type": "noul", "noul": body["state"].(map[string]Value)["probability"]}
		}
		_ = json.NewEncoder(w).Encode(map[string]Value{"model": body["model"], "answers": answers, "usage": map[string]Value{"input_tokens": 1, "output_tokens": 1}})
	}))
	defer server.Close()
	client := Typesafe(map[string]Value{"api_key": "test", "base_url": server.URL, "retry": map[string]Value{"maxRetries": 1, "initialDelayMs": 1}})
	models, err := client.ListModels(context.Background(), nil)
	if err != nil || len(models) != 1 || calls.Load() != 2 {
		t.Fatalf("model discovery/retry: %v %v", models, err)
	}
	var wg sync.WaitGroup
	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			key := fmt.Sprintf("question%d", i)
			probability := float64(i) / 10
			result, err := client.SystemOne(context.Background(), TypesafeRequest{State: map[string]Value{"probability": probability}, Questions: map[string]TypesafeQuestion{key: {Type: "noul"}}}, nil)
			if err != nil || result.Answers[key].Noul != probability {
				t.Errorf("request isolation: %v %v", result, err)
			}
		}(i)
	}
	wg.Wait()
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	done := make(chan error, 1)
	go func() {
		_, err := client.SystemOne(ctx, TypesafeRequest{State: "stall", Questions: map[string]TypesafeQuestion{"flag": {Type: "noul"}}}, nil)
		done <- err
	}()
	select {
	case <-started:
	case <-time.After(2 * time.Second):
		t.Fatal("native call did not start")
	}
	cancel()
	parityWaitResult(t, done, "abort")
}

// Exercise the built-in socket, including concurrent requests and remote close.
func TestMCPWebSocketNativeRoundTrip(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		socket, err := websocket.Accept(w, r, nil)
		if err != nil {
			t.Error(err)
			return
		}
		defer socket.CloseNow()
		for {
			_, data, err := socket.Read(r.Context())
			if err != nil {
				return
			}
			var request map[string]Value
			if err = json.Unmarshal(data, &request); err != nil {
				t.Error(err)
				return
			}
			if request["method"] == "close" {
				return
			}
			response, _ := json.Marshal(map[string]Value{"jsonrpc": "2.0", "id": request["id"], "result": request["id"]})
			if err = socket.Write(r.Context(), websocket.MessageText, response); err != nil {
				return
			}
		}
	}))
	defer server.Close()
	transport := NewAxMCPWebSocketTransport("ws"+strings.TrimPrefix(server.URL, "http"), nil)
	defer transport.Close()
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	var wg sync.WaitGroup
	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			response, err := transport.SendWithContext(ctx, map[string]Value{"id": i, "method": "ping"}, nil)
			if err != nil || response["result"] != float64(i) {
				t.Errorf("native socket response: %v %v", response, err)
			}
		}(i)
	}
	wg.Wait()
	if _, err := transport.SendWithContext(ctx, map[string]Value{"id": 11, "method": "close"}, nil); err == nil {
		t.Fatal("remote close did not reject pending request")
	}
	transport.mu.Lock()
	defer transport.mu.Unlock()
	if len(transport.pending) != 0 {
		t.Fatal("native socket close retained pending work")
	}
}

func TestTypesafeNestedBalancerValidation(t *testing.T) {
	typed := NewAI("typesafe", map[string]Value{"api_key": "test", "models": []Value{}}).(AxAIService)
	only, err := NewAxBalancer([]AxAIService{typed}, nil)
	if err != nil {
		t.Fatal(err)
	}
	prose := map[string]Value{"chat_prompt": []Value{map[string]Value{"role": "user", "content": "reply"}}}
	if only.ValidateChatRequest(prose) == nil {
		t.Fatal("nested Typesafe accepted prose")
	}
	var supported map[string]Value
	if err = json.Unmarshal([]byte(`{"chat_prompt":[{"role":"user","content":"outage"}],"response_format":{"type":"json_schema","schema":{"name":"decision","schema":{"type":"object","properties":{"urgent":{"type":"boolean"}},"required":["urgent"]}}}}`), &supported); err != nil {
		t.Fatal(err)
	}
	if err = only.ValidateChatRequest(supported); err != nil {
		t.Fatal(err)
	}
	mixed, err := NewAxBalancer([]AxAIService{only, NewAI("openai", map[string]Value{"api_key": "test", "models": []Value{}}).(AxAIService)}, nil)
	if err != nil {
		t.Fatal(err)
	}
	candidates, err := mixed.candidateServices(prose)
	if err != nil || len(candidates) != 1 {
		t.Fatalf("nested selection: %v %v", candidates, err)
	}
}
